staking/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+#![no_std]
+
+use codec::{Decode, Encode};
+use ft_io::*;
+use gstd::{debug, exec, msg, prelude::*, ActorId};
+use scale_info::TypeInfo;
+use staking_io::*;
+
+#[derive(Debug, Default, Encode, Decode, TypeInfo)]
+struct Staking {
+    staking_token_address: ActorId,
+    reward_token_address: Option<ActorId>,
+    tokens_per_stake: u128,
+    total_staked: u128,
+    distribution_time: u64,
+    produced_time: u64,
+    reward_produced: u128,
+    stakers: BTreeMap<ActorId, Staker>,
+}
+
+static mut STAKING: Option<Staking> = None;
+
+impl Staking {
+    /// Transfers `amount` tokens from `sender` account to `recipient` account.
+    /// Arguments:
+    /// * `from`: sender account
+    /// * `to`: recipient account
+    /// * `amount`: amount of tokens
+    async fn transfer_tokens(
+        &mut self,
+        token_address: &ActorId,
+        from: &ActorId,
+        to: &ActorId,
+        amount_tokens: u128,
+    ) {
+        let _transfer_response: FTEvent = msg::send_and_wait_for_reply(
+            *token_address,
+            FTAction::Transfer {
+                from: *from,
+                to: *to,
+                amount: amount_tokens,
+            },
+            0,
+        )
+        .unwrap()
+        .await
+        .expect("Error in transfer");
+    }
+
+    /// Calculates the reward produced so far
+    fn produced(&mut self) -> u128 {
+        debug!("timestamp: {}", exec::block_timestamp());
+        self.reward_produced
+            + (exec::block_timestamp() - self.produced_time) as u128
+                / self.distribution_time as u128
+    }
+
+    /// Updates the reward produced so far and calculates tokens per stake
+    fn update_reward(&mut self) {
+        let reward_produced_at_now = self.produced();
+
+        debug!("reward_produced_at_now: {}", reward_produced_at_now);
+
+        if reward_produced_at_now > self.reward_produced {
+            let produced_new = reward_produced_at_now - self.reward_produced;
+
+            if self.total_staked > 0 {
+                self.tokens_per_stake = self
+                    .tokens_per_stake
+                    .saturating_add(produced_new * (1 << 20) / self.total_staked);
```

Comment:
`a * (1 << n) = a << n`
Also, why 20? It is better to use constants for such kinds of magic numbers.
And this operation is not protected against overflow.